### PR TITLE
Consider smoothness in all profiles that use surface

### LINKED ISSUE #4935: Consider smoothness in default bicycle profile
Please consider supporting the smoothness tag in the default bicycle profile (which I assume is the one deployed at https://www.openstreetmap.org - do you know if this is the case?).

At the moment, the profile seems to completely ignore smoothness, because even ways with `smoothness=impassable` are routed over.

Exactly which smoothness values should get penalties and which ones should be avoided completely depends on what kind of bike the profile is aimed at (road bikes, general purpose / hybrid / city bikes, mountain bikes). See [Wiki](https://wiki.openstreetmap.org/wiki/Key:smoothness#Values).

Examples of ways with `smoothness=impassable`:

* Way [4905783](https://www.openstreetmap.org/way/4905783), and [route](https://www.openstreetmap.org/directions?engine=fossgis_valhalla_bicycle&route=55.92495%2C-3.42926%3B55.92070%2C-3.43475)
* Way [829106990](https://www.openstreetmap.org/way/829106990), and [route](https://www.openstreetmap.org/directions?engine=fossgis_valhalla_bicycle&route=57.295280%2C-3.414090%3B57.296900%2C-3.405030)

What I don't really understand is why the following way is avoided: way [1087976687](https://www.openstreetmap.org/way/1087976687), and [route](https://www.openstreetmap.org/directions?engine=fossgis_valhalla_bicycle&route=47.273668%2C11.369600%3B47.275451%2C11.368243)

It has very similar tags to the other two.

[Overpass query to find more](https://overpass-ultra.us/#m=16.33/47.2745/11.3712&q=PQKgUAKgFglgzgAngghgOwQUwB4oLYAOANpggPIBumATgSnIgI4CuNAngHSTVtIAuCAPbMBAI14FqmBjDQBzBHyikASswyiRfQRhSjBVAIRgQwMAG1R+7AC4A3nauDsAXxcBdANxgAFGADuKGzmsHJQgWwAvHRK7uZweIKCSmjScJEwhPRweiRxhqIwAMZsRXnmhlSwZZj5idrUAPpVxeWGOUWNcEUoed4AlN7CAkWYaHw0nkIiCHKYgnieQA)

### PR DESCRIPTION
**Pull request by gknisely**
# Issue

It has been determined that we don’t consider using the smoothness tag for setting the access and/or surface. https://wiki.openstreetmap.org/wiki/Key:smoothness

We should process the smoothness tag, but we should possibly consider turning off access based on this tag.  We need to gather some test cases from real world data to determine how to process.

We need to look into using smoothness to apply/update to the DirectedEdge surface setting as it seems to map to DE surface tags. These can easily be used when no OSM surface tag exists (but smoothness tag exists). More work would be needed when both surface and smoothness tags exist.

Smoothness to surface mapping:

smoothness=excellent and smoothness=good = kPavedSmooth

smoothness=intermediate = kPavedRough

smoothness=bad = kCompacted

smoothness=very_bad = kDirt

smoothness=horrible = kGravel

smoothness=very_horrible = kPath

## Tasklist

 - [x] Add tests
 - [x] fixes #4935
 ~- [ ] Update the docs with any new request parameters or changes to behavior described~
 - [x] Update the [changelog](CHANGELOG.md)
 - [x] If you made changes to the lua files, update the [taginfo](taginfo.json) too.

## Requirements / Relations

 Link any requirements here. Other pull requests this PR is based on?

### CODE REVIEW

**Comment by gknisely** [comment]
@kevinkreiser @chrstnbwnkl here are some numbers. I looked into `surface=impassable` and most users prefer using `smoothness=Impassable` vs using `surface=impassable`. In fact there are only 7 ways with surface=impassable

old data:
Edge Count = 808215892, Edge Length = 348749.468750

Paved Smooth Count = 610236062, Paved Smooth Length = 1981234.250000
Paved Count = 12291344,Paved Length = 666731.812500
Paved Rough Count = 2790388, Paved Rough Length = 554595.125000
Compacted Count = 43501636, Compacted Length = 871664.937500
Dirt Count = 72437224, Dirt Length = 3536336.500000
Gravel Count = 60935892, Gravel Length = 3877911.500000
Path Count = 6023346,Path Length = 1745177.625000
Impassable Count = 0, Impassable Length = 0.000000

New data:
Edge Count = 808215892, Edge Length = 348749.468750

Paved Smooth Count = 610103036, Paved Smooth Length = 1930987.500000
Paved Count = 12277928,Paved Length = 665394.250000
Paved Rough Count = 2871832, Paved Rough Length = 574564.562500
Compacted Count = 43573994, Compacted Length = 904274.250000
Dirt Count = 72374882, Dirt Length = 3505532.750000
Gravel Count = 60967256, Gravel Length = 3895266.750000
Path Count = 6035682,Path Length = 1753392.500000
Impassable Count = 11282, Impassable Length = 4239.082031

So only 4,239.082031 km of roads will get tossed. Most of these are tracks and paths. I think we are good to go. Let me know if you have questions.

File: test/gurka/test_smoothness.cc [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+#if !defined(VALHALLA_SOURCE_DIR)
+#define VALHALLA_SOURCE_DIR
+#endif
+
+using namespace valhalla;
+using namespace valhalla::baldr;
+using namespace valhalla::gurka;
+using namespace valhalla::mjolnir;
+
+const std::unordered_map<std::string, std::string> build_config{
```
Comment (kevinkreiser) [inline test/gurka/test_smoothness.cc:16]: are admins actually needed?

File: src/mjolnir/pbfgraphparser.cc [hunk 2]
```diff
@@ -1481,6 +1481,28 @@ struct graph_callback : public OSMPBF::Callback {
...
+        has_surface_ = true;
+        if (tag_.second == "excellent" || tag_.second == "good") {
+          way_.set_surface(Surface::kPavedSmooth);
+        } else if (tag_.second == "intermediate") {
+          way_.set_surface(Surface::kPavedRough);
+        } else if (tag_.second == "bad") {
+          way_.set_surface(Surface::kCompacted);
+        } else if (tag_.second == "very_bad") {
+          way_.set_surface(Surface::kDirt);
+        } else if (tag_.second == "horrible") {
```
Comment (kevinkreiser) [inline src/mjolnir/pbfgraphparser.cc:1497]: hahhah the values for this are pretty funny!

File: src/mjolnir/pbfgraphparser.cc [hunk 3]
```diff
@@ -1481,6 +1481,28 @@ struct graph_callback : public OSMPBF::Callback {
         has_surface_ = false;
       }
     };
+
+    // surface and tracktype tag should win over smoothness.
+    tag_handlers_["smoothness"] = [this]() {
+      if (!has_surface_tag_ && !has_tracktype_tag_) {
+        has_surface_ = true;
```
[the remark this answers is not in the document]
Comment (kevinkreiser) [reply src/mjolnir/pbfgraphparser.cc:1488]: oh i see we handle it right above this! cool

**Comment by kevinkreiser** [comment]
looks good to me, would be nice to put the tests in with some other test file to reduce the total number of these

**Comment by gknisely** [comment]
> looks good to me, would be nice to put the tests in with some other test file to reduce the total number of these

ok will do

File: test/gurka/test_exclude_unpaved_smoothness.cc [hunk 4]
```diff
@@ -0,0 +1,395 @@
...
+      } else if (smoothness_tag.second == "bad") {
+        EXPECT_EQ(AB_edge->surface(), Surface::kCompacted);
+        EXPECT_EQ(BA_edge->surface(), Surface::kCompacted);
+      } else if (smoothness_tag.second == "very_bad") {
+        EXPECT_EQ(AB_edge->surface(), Surface::kDirt);
+        EXPECT_EQ(BA_edge->surface(), Surface::kDirt);
+      } else if (smoothness_tag.second == "horrible") {
+        EXPECT_EQ(AB_edge->surface(), Surface::kGravel);
+        EXPECT_EQ(BA_edge->surface(), Surface::kGravel);
+      } else if (smoothness_tag.second == "very_horrible") {
```
Comment (chrstnbwnkl) [inline test/gurka/test_exclude_unpaved_smoothness.cc:336]: I wonder how people came up with "very_horrible" haha

**Review by chrstnbwnkl — APPROVED** [review]

**Pull request merged**